analyzers/src/SonarAnalyzer.Common/Extensions/LocationExtensions.cs

```diff
@@ -18,12 +18,88 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using Microsoft.CodeAnalysis.CSharp;
+
 namespace SonarAnalyzer.Extensions;
 
 public static class LocationExtensions
 {
     public static FileLinePositionSpan GetMappedLineSpanIfAvailable(this Location location) =>
-        GeneratedCodeRecognizer.IsRazorGeneratedFile(location.SourceTree)
-            ? location.GetMappedLineSpan()
-            : location.GetLineSpan();
+        GeneratedCodeRecognizer.IsRazorGeneratedFile(location.SourceTree) ? location.GetMappedLineSpan() : location.GetLineSpan();
+
+    public static FileLinePositionSpan GetMappedLineSpanIfAvailable(this Location location, SyntaxToken token)
+    {
+        var node = token.Parent;
+        if (GeneratedCodeRecognizer.IsRazorGeneratedFile(location.SourceTree))
+        {
+            var mappedLocation = location.GetMappedLineSpan();
+            if (mappedLocation.HasMappedPath)
+            {
+                var lineDirective = FindLineDirecitive(node);
+                if (LineSpanDirectiveTriviaSyntaxWrapper.IsInstance(lineDirective)
+                    && (LineSpanDirectiveTriviaSyntaxWrapper)lineDirective is var lineSpanDirective
+                    && lineSpanDirective.CharacterOffset.ValueText is var stringValue
+                    && int.TryParse(stringValue, out var numericValue)
+                    && numericValue >= location.GetLineSpan().Span.End.Character)
+                {
+                    return location.GetLineSpan();
+                }
+            }
+            return mappedLocation;
+        }
+        return location.GetLineSpan();
+    }
+
+    private static SyntaxNode FindLineDirecitive(SyntaxNode node)
+    {
+        while (node != null)
```

Comment:
With this approach, we will have lots of iterations since we will go up and down in the tree multiple times. I wander how this compares in performance with a single visit for the tree that will build a cache of the directives that we can later use when doing the mapping.

Reply:
Discussed offline, and the algorithm has changed so I am resolving this.